File: inorder-successor-in-bst/inorder-successor-in-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode inorderSuccessor(TreeNode root, TreeNode p) {
        if (root == null)return null;
        List<TreeNode> result = new ArrayList<>();
        helper(root, result);
        for(int i=0; i< result.size(); i++){
            if(result.get(i) ==p) {
                if (i+1< result.size()) {
                    return result.get(i+1);
                }
            }
        }
        return null;
    }
    public void helper(TreeNode root, List<TreeNode> result){
        if(root == null)return;
        helper(root.left, result);
        result.add(root);
        helper(root.right, result);
    }
}

File: inorder-successor-in-bst/inorder-successor-in-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderSuccessor(self, root: TreeNode, p: TreeNode) -> Optional[TreeNode]:
        successor = None
        while root:
            if root.val <= p.val:
                root = root.right
            else: 
                successor = root;
                root = root.left
        return successor
    
#time compelxity: O(n) in general and O(logn) for a balanced BST
#space complexity:O(1) since we don't use recursion